File: sim-src/environment.js




//environment.js

class environment {

    populations = {};

    env_data = {};

    stop;

    runtimes = 0;

    printPop(){
        console.log("this.populations: \n")
        console.log(this.populations);

        for (let key in this.populations) {
            console.log("key = " + key);
            console.log(" \n(this.populations)[key] \n")
            console.log(this.populations[key]);
            console.log(this.populations[key], " = ", this.populations[key].print());            
        }
    }

    pop_setup(){
        this.populations["birds"] = new population(500, 450, 100);
        this.populations["worms"] = new population(2500, 3000, 400);
        this.populations["birds"].dependencies = [[[[this.populations["worms"],1]],1]];

        //this function is no longer called, the  "setup" function in this class is currently what is run...
        //and sets up the populations based on user input
    }

    run(){


        this.runtimes++
        if (this.stop == undefined){
            return;
        }

        var iteration = 0;


        for (const population in this.populations) {
            this.env_data[population] = [];
        }

        var plz_break = false;
  

        while(true){




            var New = this.populations;
            // ^ creating new copy of population values to store new values...
            // isolated from old ones that the simulation is running on

            var extinct = [];

            for (const population in New) {

                New[population].increment();

                //console.log(`${population}: ${New[population].population_curve.num}`);
                
                if (New[population].population_curve.num <= 0){
                    extinct.push(population);
                }

                this.env_data[population] = this.populations[population].data;
            }

            iteration++;


            if (iteration >= this.stop){
                plz_break = true;
            }

            for (const extinct_index in extinct){
                var ex = extinct[extinct_index];
                delete New[ex];
                plz_break = true;
            }
            this.populations = New;
            

            if (plz_break){
                return;
            }            
        }


        /* for (const population in this.populations) {
            this.env_data[population] = this.populations[population].data;
        } */
    }
}

File: sim-src/environment.py
import time
from population import *



class environment():

    def __init__(self):
        self.populations = {}
    
    def printPop(self):
        print(self.populations)
        for i in self.populations.keys():
            print(i, "=", self.populations[i].print( ))  


    def setup(self):

        print("population number: ")

        populationsNum = int(input())
        for _ in range(populationsNum):
            print(_, "nth population: name, number, mean, standard deviation")
            nm = str(input())
            #name ^
            self.populations[nm] = population(int(input()),int(input()),int(input()))
            print(nm, "-- number:", self.populations[nm].population.num, "mean:", self.populations[nm].population.mean, "standard dev:", self.populations[nm].population.stddev)
        
        print("dependencies: ")

        key_list = list(self.populations.keys())
        val_list = list(self.populations.values())

        for i in key_list:
            pop = self.populations[i]
            print("number of elements for:", i)
            elementNum = int(input())
            listing = []
            for element in range(elementNum):
                print("element importance: ")
                imp = int(input())
                newListing = [[],imp]
                print("number of suppliers:")
                supplierNum = int(input())
                for supplier in range(supplierNum):
                    print("supplier importance, name:")
                    supplierImp = int(input())
                    supplierName = str(input())
                    newListing[0].append([self.populations[supplierName], supplierImp])
                listing.append(newListing)
                pop.dependencies = listing

        
        print("setup done!")
        self.printPop()
        def timedown(n):
            print("running in", n)
            time.sleep(1)
        timedown(3)
        timedown(2)
        timedown(1)
        print("starting")



    def pop_setup(self):
        self.populations["birds"] = population(500, 450, 100)
        self.populations["worms"] = population(2500, 3000, 400)
        self.populations["birds"].dependencies = [[[[self.populations["worms"],5]],5]]
        self.printPop()
        #this function is no longer called, the  "setup" function in this class is currently what is run...
        #and sets up the populations based on user input
    
    def run(self):
        while True:

            new = self.populations
            # ^ creating new copy of population values to store new values...
            # isolated from old ones that the simulation is running on

            extinct = [] # list for extinct species

            for population in new:
                new[population].increment()
                print(population,  ": " , new[population].population.num)
                if new[population].population.num <= 0:
                    print(population, "went extinct")
                    extinct.append(population)

            for ex in extinct:
                new.pop(ex)

            self.populations = new
            time.sleep(1)
                
